File: kotlin/src/main/kotlin/com/edgeveda/sdk/WhisperSession.kt
package com.edgeveda.sdk

import com.edgeveda.sdk.internal.NativeBridge
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.util.concurrent.atomic.AtomicBoolean

/**
 * WhisperSession - Persistent speech-to-text (STT) inference manager
 *
 * Maintains a persistent native Whisper context that is loaded once and reused
 * for all audio transcriptions. Supports automatic language detection, translation,
 * and timestamp-accurate segment extraction.
 *
 * Pattern mirrors VisionWorker, adapted for audio transcription with Whisper models.
 *
 * Usage:
 * ```kotlin
 * val session = WhisperSession()
 * session.initialize(WhisperConfig(modelPath = "/path/to/ggml-model.bin"))
 * 
 * // Transcribe audio (PCM 16kHz mono float32, range [-1.0, 1.0])
 * val result = session.transcribe(pcmSamples)
 * 
 * result.segments.forEach { segment ->
 *     println("[${segment.startMs}ms - ${segment.endMs}ms] ${segment.text}")
 * }
 * 
 * // Translate to English
 * val translated = session.transcribe(
 *     pcmSamples,
 *     WhisperTranscribeParams(language = "auto", translate = true)
 * )
 * 
 * // Cleanup
 * session.cleanup()
 * ```
 */
class WhisperSession {
    private val _isInitialized = AtomicBoolean(false)
    @Volatile private var backend = ""

    /**
     * Whether the whisper context is initialized and ready
     */
    val isInitialized: Boolean
        get() = _isInitialized.get()

    /**
     * Backend name (e.g., "Metal", "CPU")
     */
    val backendName: String
        get() = backend

    /**
     * Initialize the whisper context with STT model
     *
     * Loads the Whisper model once. Subsequent transcription requests reuse
     * this context without reloading.
     *
     * @param config Whisper configuration
     * @return Backend name (e.g., "Metal", "CPU")
     * @throws EdgeVedaException if initialization fails
     */
    suspend fun initialize(config: WhisperConfig): String = withContext(Dispatchers.Default) {
        if (!_isInitialized.compareAndSet(false, true)) {
            throw EdgeVedaException.InvalidConfiguration(
                "WhisperSession already initialized"
            )
        }

        try {
            val configJson = JSONObject().apply {
                put("modelPath", config.modelPath)
                put("numThreads", config.numThreads)
                put("useGpu", config.useGpu)
            }.toString()

            backend = NativeBridge.initWhisper(configJson)
            backend
        } catch (e: EdgeVedaException) {
            _isInitialized.set(false) // rollback on failure
            throw e
        } catch (e: Exception) {
            _isInitialized.set(false) // rollback on failure
            throw EdgeVedaException.ModelLoadError(
                "Failed to initialize whisper context: ${e.message}",
                e
            )
        }
    }

    /**
     * Transcribe PCM audio samples to text with timing segments
     *
     * Audio must be PCM 16kHz mono float32 with samples in range [-1.0, 1.0].
     * Use WhisperTranscribeParams to control language detection, translation,
     * and threading.
     *
     * @param pcmSamples PCM audio data (16kHz, mono, float32, range [-1.0, 1.0])
     * @param params Optional transcription parameters
     * @return Transcription result with text segments and timestamps
     * @throws EdgeVedaException if transcription fails
     */
    suspend fun transcribe(
        pcmSamples: FloatArray,
        params: WhisperTranscribeParams = WhisperTranscribeParams()
    ): WhisperResult = withContext(Dispatchers.Default) {
        if (!_isInitialized.get()) {
            throw EdgeVedaException.InvalidConfiguration(
                "WhisperSession not initialized. Call initialize() first."
            )
        }

        try {
            val paramsJson = JSONObject().apply {
                put("language", params.language)
                put("translate", params.translate)
                put("nThreads", params.nThreads)
            }.toString()

            val resultJson = NativeBridge.transcribeAudio(pcmSamples, paramsJson)
            val parsed = JSONObject(resultJson)

            // Parse segments array
            val segmentsArray = parsed.getJSONArray("segments")
            val segments = mutableListOf<WhisperSegment>()

            for (i in 0 until segmentsArray.length()) {
                val segmentObj = segmentsArray.getJSONObject(i)
                segments.add(
                    WhisperSegment(
                        text = segmentObj.getString("text"),
                        startMs = segmentObj.getLong("start_ms"),
                        endMs = segmentObj.getLong("end_ms")
                    )
                )
            }

            WhisperResult(
                segments = segments,
                fullText = segments.joinToString(" ") { it.text.trim() }
            )
        } catch (e: EdgeVedaException) {
            throw e
        } catch (e: Exception) {
            throw EdgeVedaException.GenerationError(
                "Whisper transcription failed: ${e.message}",
                e
            )
        }
    }

    /**
     * Clean up and free native whisper resources
     *
     * Frees the native whisper context (model). The session cannot be used
     * after cleanup unless initialize() is called again.
     *
     * @throws EdgeVedaException if cleanup fails
     */
    suspend fun cleanup() = withContext(Dispatchers.Default) {
        if (!_isInitialized.get()) {
            return@withContext
        }

        try {
            NativeBridge.freeWhisper()
            _isInitialized.set(false)
            backend = ""
        } catch (e: Exception) {
            throw EdgeVedaException.NativeError(
                "Failed to cleanup whisper context: ${e.message}",
                e
            )
        }
    }
}

/**
 * Configuration for WhisperSession initialization.
 *
 * @property modelPath Path to the Whisper GGML model file
 * @property numThreads Number of CPU threads (0 = auto-detect, default: 4)
 * @property useGpu Enable GPU acceleration if available (default: true)
 */
data class WhisperConfig(
    val modelPath: String,
    val numThreads: Int = 4,
    val useGpu: Boolean = true
) {
    init {
        require(modelPath.isNotEmpty()) { "modelPath cannot be empty" }
        require(numThreads >= 0) { "numThreads must be >= 0" }
    }
}

/**
 * Parameters for audio transcription.
 *
 * @property language Language code (e.g., "en", "es", "fr", "auto" for detection, default: "en")
 * @property translate Translate to English if source language is not English (default: false)
 * @property nThreads Number of threads for transcription (0 = auto-detect, default: 4)
 */
data class WhisperTranscribeParams(
    val language: String = "en",
    val translate: Boolean = false,
    val nThreads: Int = 4
) {
    init {
        require(language.isNotEmpty()) { "language cannot be empty" }
        require(nThreads >= 0) { "nThreads must be >= 0" }
    }

    companion object {
        /**
         * Default params for automatic language detection
         */
        fun autoDetect(): WhisperTranscribeParams = WhisperTranscribeParams(
            language = "auto",
            translate = false
        )

        /**
         * Params for transcription + translation to English
         */
        fun autoTranslate(): WhisperTranscribeParams = WhisperTranscribeParams(
            language = "auto",
            translate = true
        )

        /**
         * Params for fast CPU-only transcription
         */
        fun fast(): WhisperTranscribeParams = WhisperTranscribeParams(
            language = "en",
            translate = false,
            nThreads = 2
        )
    }
}

/**
 * Result from Whisper transcription.
 *
 * @property segments List of text segments with timestamps
 * @property fullText Complete transcription text (all segments joined)
 */
data class WhisperResult(
    val segments: List<WhisperSegment>,
    val fullText: String
) {
    /**
     * Total duration of transcribed audio in milliseconds
     */
    val durationMs: Long
        get() = segments.lastOrNull()?.endMs ?: 0L

    /**
     * Number of segments in the transcription
     */
    val segmentCount: Int
        get() = segments.size
}

/**
 * A single transcribed text segment with timing information.
 *
 * @property text Transcribed text for this segment
 * @property startMs Start time in milliseconds
 * @property endMs End time in milliseconds
 */
data class WhisperSegment(
    val text: String,
    val startMs: Long,
    val endMs: Long
) {
    /**
     * Duration of this segment in milliseconds
     */
    val durationMs: Long
        get() = endMs - startMs

    /**
     * Start time in seconds (for convenience)
     */
    val startSeconds: Double
        get() = startMs / 1000.0

    /**
     * End time in seconds (for convenience)
     */
    val endSeconds: Double
        get() = endMs / 1000.0
}

File: react-native/src/WhisperSession.ts
/**
 * WhisperSession - Persistent speech-to-text inference context
 *
 * Manages a persistent native Whisper context loaded once and reused
 * for subsequent transcriptions. Mirrors the VisionWorker.ts pattern
 * adapted for audio inference.
 *
 * Pattern mirrors Swift's WhisperSession and Flutter's WhisperSession,
 * adapted for React Native's TurboModule architecture.
 *
 * Usage:
 * 1. Create WhisperSession instance with modelPath
 * 2. Call initialize() to load the Whisper model
 * 3. Call transcribe() with Float32Array PCM samples
 * 4. Call cleanup() when done
 */

import NativeEdgeVeda from './NativeEdgeVeda';
import type { WhisperConfig, WhisperParams, WhisperResult } from './types';
import { EdgeVedaError, EdgeVedaErrorCode } from './types';

// Base64 encoding lookup table
const BASE64_CHARS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';

/**
 * WhisperSession - Manages persistent Whisper STT inference context
 *
 * Handles native Whisper context lifecycle and audio transcription.
 * Float32Array PCM data is base64-encoded before transfer to the
 * native layer (same pattern as VisionWorker's RGB frame encoding).
 */
export class WhisperSession {
  /** Absolute path to the Whisper GGUF model file */
  readonly modelPath: string;

  private _isInitialized = false;
  private _backend = '';

  constructor(modelPath: string) {
    this.modelPath = modelPath;
  }

  /**
   * Whether the Whisper context is initialized and ready
   */
  get isInitialized(): boolean {
    return this._isInitialized;
  }

  /**
   * Backend used for inference (e.g., "Metal", "CUDA", "CPU")
   */
  get backend(): string {
    return this._backend;
  }

  /**
   * Initialize the Whisper context
   *
   * Loads the Whisper model once. Subsequent transcribe() calls
   * reuse this context without reloading.
   *
   * @param config - Optional Whisper configuration (overrides modelPath if provided)
   * @returns Backend name (e.g., "Metal", "CPU")
   */
  async initialize(config?: WhisperConfig): Promise<string> {
    if (this._isInitialized) {
      throw new EdgeVedaError(
        EdgeVedaErrorCode.INVALID_PARAMETER,
        'WhisperSession already initialized'
      );
    }

    try {
      const modelPath = config?.modelPath ?? this.modelPath;
      const configJson = JSON.stringify({
        numThreads: config?.numThreads ?? 4,
        useGpu: config?.useGpu ?? false,
      });

      this._backend = await NativeEdgeVeda.initWhisper(modelPath, configJson);
      this._isInitialized = true;
      return this._backend;
    } catch (error) {
      throw new EdgeVedaError(
        EdgeVedaErrorCode.MODEL_LOAD_FAILED,
        'Failed to initialize Whisper context',
        error instanceof Error ? error.message : String(error)
      );
    }
  }

  /**
   * Transcribe audio PCM samples to text
   *
   * @param pcmSamples - 16 kHz mono Float32Array PCM audio data
   * @param params - Optional transcription parameters
   * @returns Transcription result with segments and full text
   */
  async transcribe(
    pcmSamples: Float32Array,
    params?: WhisperParams
  ): Promise<WhisperResult> {
    if (!this._isInitialized) {
      throw new EdgeVedaError(
        EdgeVedaErrorCode.MODEL_NOT_LOADED,
        'WhisperSession not initialized. Call initialize() first.'
      );
    }

    try {
      // Encode Float32Array as Base64 for TurboModule transfer
      const pcmBase64 = this._float32ArrayToBase64(pcmSamples);

      const paramsJson = JSON.stringify({
        language: params?.language ?? 'en',
        translate: params?.translate ?? false,
        maxTokens: params?.maxTokens ?? 0,
      });

      const resultJson = await NativeEdgeVeda.transcribeAudio(
        pcmBase64,
        pcmSamples.length,
        paramsJson
      );

      return JSON.parse(resultJson) as WhisperResult;
    } catch (error) {
      throw new EdgeVedaError(
        EdgeVedaErrorCode.GENERATION_FAILED,
        'Whisper transcription failed',
        error instanceof Error ? error.message : String(error)
      );
    }
  }

  /**
   * Clean up and free native Whisper resources
   *
   * Frees the native Whisper context. The session cannot be used
   * after cleanup unless initialize() is called again.
   */
  async cleanup(): Promise<void> {
    if (!this._isInitialized) {
      return;
    }

    try {
      await NativeEdgeVeda.freeWhisper();
      this._isInitialized = false;
      this._backend = '';
    } catch (error) {
      throw new EdgeVedaError(
        EdgeVedaErrorCode.UNKNOWN_ERROR,
        'Failed to cleanup Whisper context',
        error instanceof Error ? error.message : String(error)
      );
    }
  }

  /**
   * Convert Float32Array to Base64 string for TurboModule transfer
   *
   * Reinterprets the Float32 bytes as raw bytes before encoding,
   * preserving the full IEEE 754 float precision on the native side.
   */
  private _float32ArrayToBase64(samples: Float32Array): string {
    // Reinterpret float bytes as raw uint8 bytes
    const bytes = new Uint8Array(samples.buffer, samples.byteOffset, samples.byteLength);
    let result = '';
    const len = bytes.length;

    // Process 3 bytes at a time into 4 Base64 characters
    for (let i = 0; i < len; i += 3) {
      const byte1 = bytes[i];
      const byte2 = i + 1 < len ? bytes[i + 1] : 0;
      const byte3 = i + 2 < len ? bytes[i + 2] : 0;

      const encoded1 = byte1 >> 2;
      const encoded2 = ((byte1 & 0x03) << 4) | (byte2 >> 4);
      const encoded3 = ((byte2 & 0x0f) << 2) | (byte3 >> 6);
      const encoded4 = byte3 & 0x3f;

      result += BASE64_CHARS[encoded1];
      result += BASE64_CHARS[encoded2];
      result += i + 1 < len ? BASE64_CHARS[encoded3] : '=';
      result += i + 2 < len ? BASE64_CHARS[encoded4] : '=';
    }

    return result;
  }
}
